### PR TITLE
ifp: Allow constraining row count to be odd or even

### PR DESCRIPTION
**Pull request by gatecat**
This is an alternative solution to #5429 - for gf130 we need an even row count.

I haven't considered the hybrid row case, as that's not needed for us and it seems like supporting it there would introduce a various edge cases that would need to be carefully considered.

### CODE REVIEW

File: src/ifp/README.md [hunk 1]
```diff
@@ -72,6 +73,7 @@ initialize_floorplan
 | `-site` | Site name. |
 | `-die_area` | Die area coordinates in microns (lower left x/y and upper right x/y coordinates). |
 | `-core_area` | Core area coordinates in microns (lower left x/y and upper right x/y coordinates). |
+| `-row_parity` | Snap to either an odd (`ODD`) or even (`EVEN`) number of rows. |
```
Comment (maliberty) [inline src/ifp/README.md:76]: Indicate the default (presumably NONE).

File: src/ifp/src/InitFloorplan.i [hunk 2]
```diff
@@ -97,11 +110,12 @@ init_floorplan_core(int die_lx,
 		    int core_ux,
 		    int core_uy,
 		    odb::dbSite* site,
-		    const std::vector<odb::dbSite*>& additional_sites)
+		    const std::vector<odb::dbSite*>& additional_sites,
+        ifp::RowParity row_parity)
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.i:114]: whitespace alignment is off

File: src/ifp/src/InitFloorplan.tcl [hunk 3]
```diff
@@ -39,12 +39,13 @@ sta::define_cmd_args "initialize_floorplan" {[-utilization util]\
 					       [-die_area {lx ly ux uy}]\
 					       [-core_area {lx ly ux uy}]\
 					       [-additional_sites site_names]\
-					       [-site site_name]}
+					       [-site site_name]\
+                 [-row_parity NONE|ODD|EVEN]}
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.tcl:43]: alignment

File: src/ifp/src/InitFloorplan.cc [hunk 4]
```diff
@@ -416,7 +426,22 @@ void InitFloorplan::makeUniformRows(odb::dbSite* base_site,
...
+
+    switch (row_parity) {
+      case RowParity::NONE:
+        break;
+      case RowParity::EVEN:
+        rows_y = (rows_y / 2) * 2;
+        break;
+      case RowParity::ODD:
+        if (rows_y > 0) {
+          rows_y = ((rows_y - 1) / 2) * 2 + 1;
```
Comment (maliberty) [inline src/ifp/src/InitFloorplan.cc:439]: Fwiw this works but I find `rows_y = rows_y % 2 ? rows_y - 1 : rows_y` more obvious

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/ifp/src/InitFloorplan.i (hunk 2 above)
Comment (gatecat) [reply src/ifp/src/InitFloorplan.i]: Fixed - these files are using a mix of tabs and spaces which I guess should eventually be fixed, but I won't do that in this PR to keep the diff consistent.

File: src/ifp/src/InitFloorplan.cc (hunk 4 above)
Comment (gatecat) [reply src/ifp/src/InitFloorplan.cc]: Thanks, I agree, I've changed it.

**Review by maliberty — APPROVED** [review]

**Pull request merged**